src/main/scala/uk/co/gresearch/spark/diff/DiffComparator.scala

```diff
@@ -0,0 +1,118 @@
+package uk.co.gresearch.spark.diff
+
+import org.apache.spark.sql.Column
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.catalyst.util.DateTimeUtils._
+import org.apache.spark.sql.types.{DataType, DoubleType, FloatType}
+
+/**
+ * Abstract class for a Comparator factory of a given comparison type.
+ * @param symbols - list of symbols referencing the given comparison method
+ */
+abstract class DiffComparator(val symbols: Seq[String]) {
+  
+  def description: String
+
+  /**
+   * The Compare Expression of two other expressions (usually resolved as instance of [[BinaryExpression]]
+   * @param left - the left expression of a binary expression (left Column)
+   * @param right - the right expression of a binary expression (right Column)
+   */
+  def compare(left: Column, right: Column): Expression
```

Comment:
I wouldn't base this trait on `Expression`. User implementation can use Expression and then provide a `Column(expr)` to the trait. Don't see the benefit of not using `Column` only.